spacy_experimental/coref/coref_component.py

```diff
@@ -0,0 +1,426 @@
+
+    def rehearse(self, examples, *, sgd=None, losses=None, **config):
+        # TODO this should be added later
+        raise NotImplementedError(
+            Errors.E931.format(
+                parent="CoreferenceResolver", method="add_label", name=self.name
+            )
+        )
+
+    def add_label(self, label: str) -> int:
+        """Technically this method should be implemented from TrainablePipe,
+        but it is not relevant for the coref component.
+        """
+        raise NotImplementedError(
+            Errors.E931.format(
+                parent="CoreferenceResolver", method="add_label", name=self.name
+            )
+        )
+
+    def get_loss(
+        self,
+        examples: Iterable[Example],
+        score_matrix: Floats2d,
+        mention_idx: Ints2d,
+    ):
+        """Find the loss and gradient of loss for the batch of documents and
+        their predicted scores.
+
+        examples (Iterable[Examples]): The batch of examples.
+        score_matrix (Floats2d): Scores representing the model's predictions.
+        mention_idx (Ints2d): Mention indices.
+        RETURNS (Tuple[float, float]): The loss and the gradient.
+
+        DOCS: https://spacy.io/api/coref#get_loss
+        """
+        ops = self.model.ops
+        xp = ops.xp
+
+        examples = list(examples)
+        if len(examples) > 1:
+            # TODO handle more than one document
+            raise ValueError(
+                "Got more than one example, but only fake batching is supported."
+            )
+        example = examples[0]
+
+        clusters_by_char = get_clusters_from_doc(example.reference)
+        # convert to token clusters, and give up if necessary
+        clusters = []
+        for cluster in clusters_by_char:
+            cc = []
+            for start_char, end_char in cluster:
+                span = example.predicted.char_span(start_char, end_char)
+                if span is None:
+                    # TODO log more details
+                    raise IndexError(
+                        "Misalignment in Coref. Head token has no match in training doc."
+                    )
+                cc.append((span.start, span.end))
+            clusters.append(cc)
+
+        span_idxs = create_head_span_idxs(ops, len(example.predicted))
+        gscores = create_gold_scores(span_idxs, clusters)
+        # Note on type here. This is bools but asarray2f wants ints.
+        gscores = ops.asarray2f(gscores)  # type: ignore
+        top_gscores = xp.take_along_axis(gscores, mention_idx, axis=1)
+        # now add the placeholder
+        gold_placeholder = ~top_gscores.any(axis=1).T
+        gold_placeholder = xp.expand_dims(gold_placeholder, 1)
+        top_gscores = xp.concatenate((gold_placeholder, top_gscores), 1)
+
+        # boolean to float
+        top_gscores = ops.asarray2f(top_gscores)
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=RuntimeWarning)
```

Comment:
To build a mask over scores so that only half the matrix is used, the code here builds a matrix of 0/1 values and takes the log. Taking the log of zero results in infinite values that are used for the mask, but raises the warning that's ignored here. I'll add a comment to clarify.

Reply:
Edits here https://github.com/explosion/spacy-experimental/pull/17/commits/b736fba62f60b7c7aa3df550d00d1d5e286d1b30